fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/ArrayMatchAnyExpander.java

```diff
@@ -0,0 +1,128 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.rules.expression.rules;
+
+import org.apache.doris.analysis.IndexDef;
+import org.apache.doris.analysis.InvertedIndexUtil;
+import org.apache.doris.catalog.Index;
+import org.apache.doris.catalog.OlapTable;
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleType;
+import org.apache.doris.nereids.rules.rewrite.OneRewriteRuleFactory;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.SlotReference;
+import org.apache.doris.nereids.trees.expressions.functions.scalar.ArrayMatchAny;
+import org.apache.doris.nereids.trees.expressions.literal.Literal;
+import org.apache.doris.nereids.trees.expressions.literal.MapLiteral;
+import org.apache.doris.nereids.trees.expressions.literal.StringLiteral;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalFilter;
+import org.apache.doris.nereids.trees.plans.logical.LogicalOlapScan;
+import org.apache.doris.nereids.util.ExpressionUtils;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * ArrayMatchAnyExpander
+ *
+ *  expand array_match_any(col1, "a,b") to array_match_any(col1, "a,b", invertedIdxCtx)
+ *  if literal array , we should not expand just to go with no inverted index match
+ *   like array_match_any(["a", "b", "c", 'd'], "a,b") to array_match_any(col1, "a,b")
+ *
+ * after expand, we can evaluate it.
+ */
+public class ArrayMatchAnyExpander extends OneRewriteRuleFactory {
+    @Override
+    public Rule build() {
+        return logicalFilter(logicalOlapScan())
+                .when(filter -> ExpressionUtils.containsType(filter.getExpressions(), ArrayMatchAny.class))
+                .then(this::replace)
+                .toRule(RuleType.ARRAY_MATCH_ANY_EXPANDER);
+    }
+
+    Plan replace(LogicalFilter<LogicalOlapScan> filter) {
+        Set<Expression> sets = filter.getConjuncts().stream().map(expr -> {
+            if (expr instanceof ArrayMatchAny) {
+                return rewrite((ArrayMatchAny) expr);
+            } else {
+                return expr;
+            }
+        }).collect(ImmutableSet.toImmutableSet());
+        return filter.withConjuncts(sets);
+    }
+
+    private static Expression rewrite(ArrayMatchAny arrayMatchAny) {
```

Comment:
Is rewrite function used for generating index properties? It's too complex process each key of the properties map and not easy to maintain if new key added. Can you just copy the map instead of check each key?

Reply:
done!